src/wallet/wallet.cpp

```diff
@@ -470,6 +524,81 @@ void CWallet::AddToSpends(const uint256& wtxid)
 
     BOOST_FOREACH(const CTxIn& txin, thisTx.vin)
         AddToSpends(txin.prevout, wtxid);
+    BOOST_FOREACH(const JSDescription& jsdesc, thisTx.vjoinsplit)
+        BOOST_FOREACH(const uint256& nullifier, jsdesc.nullifiers)
+            AddToSpends(nullifier, wtxid);
+}
+
+void CWallet::IncrementNoteWitnesses(const CBlockIndex* pindex, const CBlock* pblockIn)
+{
+    {
+        LOCK(cs_wallet);
+        for (NoteWitnesses::value_type& item : mapNoteWitnesses) {
+            // Copy the witness for the previous block
+            item.second.push_front(item.second.front());
+            // Cache enough witnesses that we can expect
+            // to never reorg beyond our cache.
```

Comment:
We nevertheless need to test reorgs beyond the cache. Is that already done? If not then please file a ticket.